File: packages/sdk-java/harness-sdk-core/src/main/java/com/harness/core/CostStorage.java
package com.harness.core;

import java.sql.*;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.harness.types.TokenUsage;

/**
 * Cost storage interface for multi-level budget tracking.
 *
 * Provides storage backends for tracking user-level and global-level usage.
 *
 * Example:
 * <pre>
 * CostStorage storage = new InMemoryCostStorage();
 *
 * // Record user usage
 * UserUsage usage = storage.recordUserUsage("user-123", 1000, 500, true);
 *
 * // Check budget
 * boolean withinBudget = usage.checkBudget(config);
 *
 * // Get global usage
 * GlobalUsage global = storage.getGlobalUsage();
 * </pre>
 */
public interface CostStorage {

    /**
     * Get usage for a user.
     */
    UserUsage getUserUsage(String userId);

    /**
     * Record usage for a user.
     */
    UserUsage recordUserUsage(String userId, int inputTokens, int outputTokens, boolean request);

    /**
     * Get global usage.
     */
    GlobalUsage getGlobalUsage();

    /**
     * Record global usage.
     */
    GlobalUsage recordGlobalUsage(double costUsd, int tokens);

    /**
     * Reset daily counters.
     */
    void resetDaily();

    // -------------------------------------------------------------------------
    // User Usage
    // -------------------------------------------------------------------------

    /**
     * User-level usage statistics.
     */
    class UserUsage {
        private final String userId;
        private int dailyTokens;
        private int hourlyRequests;
        private String date;
        private int hour;

        public UserUsage(String userId) {
            this.userId = userId;
            this.date = LocalDate.now().toString();
            this.hour = LocalDateTime.now().getHour();
        }

        public UserUsage(String userId, int dailyTokens, int hourlyRequests, String date, int hour) {
            this.userId = userId;
            this.dailyTokens = dailyTokens;
            this.hourlyRequests = hourlyRequests;
            this.date = date;
            this.hour = hour;
        }

        public String userId() { return userId; }
        public int dailyTokens() { return dailyTokens; }
        public int hourlyRequests() { return hourlyRequests; }
        public String date() { return date; }
        public int hour() { return hour; }

        public void addTokens(int tokens) { this.dailyTokens += tokens; }
        public void addRequest() { this.hourlyRequests++; }

        /**
         * Check if usage exceeds budget.
         */
        public boolean checkBudget(HarnessConfig.CostControlConfig config) {
            return dailyTokens <= config.getDailyTokenLimit() &&
                   hourlyRequests <= config.getHourlyRequestLimit();
        }
    }

    // -------------------------------------------------------------------------
    // Global Usage
    // -------------------------------------------------------------------------

    /**
     * Global usage statistics.
     */
    class GlobalUsage {
        private double dailyCostUsd;
        private int dailyTokens;
        private String date;

        public GlobalUsage() {
            this.date = LocalDate.now().toString();
        }

        public GlobalUsage(double dailyCostUsd, int dailyTokens, String date) {
            this.dailyCostUsd = dailyCostUsd;
            this.dailyTokens = dailyTokens;
            this.date = date;
        }

        public double dailyCostUsd() { return dailyCostUsd; }
        public int dailyTokens() { return dailyTokens; }
        public String date() { return date; }

        public void addCost(double cost) { this.dailyCostUsd += cost; }
        public void addTokens(int tokens) { this.dailyTokens += tokens; }

        /**
         * Check if usage exceeds budget.
         */
        public boolean checkBudget(HarnessConfig.CostControlConfig config) {
            return dailyCostUsd <= config.getGlobalDailyBudgetUsd();
        }
    }
}

// -------------------------------------------------------------------------
// In-Memory Implementation
// -------------------------------------------------------------------------

/**
 * In-memory cost storage.
 *
 * Suitable for single-process applications. Data is lost on restart.
 */
class InMemoryCostStorage implements CostStorage {

    private static final Logger logger = LoggerFactory.getLogger(InMemoryCostStorage.class);

    private final Map<String, UserUsage> userUsageMap = new ConcurrentHashMap<>();
    private GlobalUsage globalUsage = new GlobalUsage();
    private String lastResetDate = "";

    private String getCurrentDate() {
        return LocalDate.now().toString();
    }

    private int getCurrentHour() {
        return LocalDateTime.now().getHour();
    }

    private void checkAndResetDaily() {
        String currentDate = getCurrentDate();
        if (!currentDate.equals(lastResetDate)) {
            resetDaily();
            lastResetDate = currentDate;
        }
    }

    @Override
    public UserUsage getUserUsage(String userId) {
        checkAndResetDaily();

        return userUsageMap.computeIfAbsent(userId, id -> new UserUsage(id));
    }

    @Override
    public UserUsage recordUserUsage(String userId, int inputTokens, int outputTokens, boolean request) {
        checkAndResetDaily();

        UserUsage usage = userUsageMap.computeIfAbsent(userId, id -> new UserUsage(id));

        // Check if hour changed
        int currentHour = getCurrentHour();
        if (currentHour != usage.hour()) {
            usage = new UserUsage(userId, usage.dailyTokens(), 0, usage.date(), currentHour);
            userUsageMap.put(userId, usage);
        }

        usage.addTokens(inputTokens + outputTokens);
        if (request) {
            usage.addRequest();
        }

        return usage;
    }

    @Override
    public GlobalUsage getGlobalUsage() {
        checkAndResetDaily();
        return globalUsage;
    }

    @Override
    public GlobalUsage recordGlobalUsage(double costUsd, int tokens) {
        checkAndResetDaily();

        globalUsage.addCost(costUsd);
        globalUsage.addTokens(tokens);

        return globalUsage;
    }

    @Override
    public void resetDaily() {
        userUsageMap.clear();
        globalUsage = new GlobalUsage();
        lastResetDate = getCurrentDate();
        logger.debug("Daily cost storage reset");
    }
}

// -------------------------------------------------------------------------
// SQLite Implementation
// -------------------------------------------------------------------------

/**
 * SQLite-based cost storage.
 *
 * Provides persistent storage for multi-process applications.
 */
class SQLiteCostStorage implements CostStorage {

    private static final Logger logger = LoggerFactory.getLogger(SQLiteCostStorage.class);

    private final String dbPath;
    private boolean initialized = false;

    public SQLiteCostStorage(String dbPath) {
        this.dbPath = dbPath;
        initDb();
    }

    public SQLiteCostStorage() {
        this(".harness/costs.db");
    }

    private void initDb() {
        if (initialized) return;

        try (Connection conn = DriverManager.getConnection("jdbc:sqlite:" + dbPath)) {
            conn.createStatement().execute("""
                CREATE TABLE IF NOT EXISTS user_usage (
                    user_id TEXT,
                    date TEXT,
                    hour INTEGER,
                    daily_tokens INTEGER DEFAULT 0,
                    hourly_requests INTEGER DEFAULT 0,
                    PRIMARY KEY (user_id, date)
                )
            """);

            conn.createStatement().execute("""
                CREATE TABLE IF NOT EXISTS global_usage (
                    date TEXT PRIMARY KEY,
                    daily_cost_usd REAL DEFAULT 0,
                    daily_tokens INTEGER DEFAULT 0
                )
            """);

            initialized = true;
        } catch (SQLException e) {
            logger.error("Failed to initialize cost storage database: {}", e.getMessage());
        }
    }

    private String getCurrentDate() {
        return LocalDate.now().toString();
    }

    private int getCurrentHour() {
        return LocalDateTime.now().getHour();
    }

    @Override
    public UserUsage getUserUsage(String userId) {
        String date = getCurrentDate();
        int hour = getCurrentHour();

        try (Connection conn = DriverManager.getConnection("jdbc:sqlite:" + dbPath)) {
            PreparedStatement stmt = conn.prepareStatement(
                "SELECT daily_tokens, hourly_requests, hour FROM user_usage WHERE user_id = ? AND date = ?"
            );
            stmt.setString(1, userId);
            stmt.setString(2, date);

            ResultSet rs = stmt.executeQuery();
            if (rs.next()) {
                int storedHour = rs.getInt("hour");
                int hourlyRequests = storedHour == hour ? rs.getInt("hourly_requests") : 0;
                return new UserUsage(userId, rs.getInt("daily_tokens"), hourlyRequests, date, hour);
            }

            return new UserUsage(userId);

        } catch (SQLException e) {
            logger.error("Failed to get user usage: {}", e.getMessage());
            return new UserUsage(userId);
        }
    }

    @Override
    public UserUsage recordUserUsage(String userId, int inputTokens, int outputTokens, boolean request) {
        String date = getCurrentDate();
        int hour = getCurrentHour();
        int tokens = inputTokens + outputTokens;

        try (Connection conn = DriverManager.getConnection("jdbc:sqlite:" + dbPath)) {
            // Check if exists
            PreparedStatement checkStmt = conn.prepareStatement(
                "SELECT daily_tokens, hourly_requests, hour FROM user_usage WHERE user_id = ? AND date = ?"
            );
            checkStmt.setString(1, userId);
            checkStmt.setString(2, date);
            ResultSet rs = checkStmt.executeQuery();

            if (rs.next()) {
                int storedHour = rs.getInt("hour");
                int hourlyRequests = storedHour == hour ? rs.getInt("hourly_requests") : 0;

                PreparedStatement updateStmt = conn.prepareStatement("""
                    UPDATE user_usage
                    SET daily_tokens = daily_tokens + ?,
                        hourly_requests = ?,
                        hour = ?
                    WHERE user_id = ? AND date = ?
                """);
                updateStmt.setInt(1, tokens);
                updateStmt.setInt(2, hourlyRequests + (request ? 1 : 0));
                updateStmt.setInt(3, hour);
                updateStmt.setString(4, userId);
                updateStmt.setString(5, date);
                updateStmt.execute();
            } else {
                PreparedStatement insertStmt = conn.prepareStatement("""
                    INSERT INTO user_usage (user_id, date, hour, daily_tokens, hourly_requests)
                    VALUES (?, ?, ?, ?, ?)
                """);
                insertStmt.setString(1, userId);
                insertStmt.setString(2, date);
                insertStmt.setInt(3, hour);
                insertStmt.setInt(4, tokens);
                insertStmt.setInt(5, request ? 1 : 0);
                insertStmt.execute();
            }

        } catch (SQLException e) {
            logger.error("Failed to record user usage: {}", e.getMessage());
        }

        return getUserUsage(userId);
    }

    @Override
    public GlobalUsage getGlobalUsage() {
        String date = getCurrentDate();

        try (Connection conn = DriverManager.getConnection("jdbc:sqlite:" + dbPath)) {
            PreparedStatement stmt = conn.prepareStatement(
                "SELECT daily_cost_usd, daily_tokens FROM global_usage WHERE date = ?"
            );
            stmt.setString(1, date);

            ResultSet rs = stmt.executeQuery();
            if (rs.next()) {
                return new GlobalUsage(rs.getDouble("daily_cost_usd"), rs.getInt("daily_tokens"), date);
            }

            return new GlobalUsage();

        } catch (SQLException e) {
            logger.error("Failed to get global usage: {}", e.getMessage());
            return new GlobalUsage();
        }
    }

    @Override
    public GlobalUsage recordGlobalUsage(double costUsd, int tokens) {
        String date = getCurrentDate();

        try (Connection conn = DriverManager.getConnection("jdbc:sqlite:" + dbPath)) {
            PreparedStatement stmt = conn.prepareStatement("""
                INSERT INTO global_usage (date, daily_cost_usd, daily_tokens)
                VALUES (?, ?, ?)
                ON CONFLICT(date) DO UPDATE SET
                    daily_cost_usd = daily_cost_usd + ?,
                    daily_tokens = daily_tokens + ?
            """);
            stmt.setString(1, date);
            stmt.setDouble(2, costUsd);
            stmt.setInt(3, tokens);
            stmt.setDouble(4, costUsd);
            stmt.setInt(5, tokens);
            stmt.execute();

        } catch (SQLException e) {
            logger.error("Failed to record global usage: {}", e.getMessage());
        }

        return getGlobalUsage();
    }

    @Override
    public void resetDaily() {
        // SQLite automatically handles new dates
    }
}

File: packages/sdk/src/harness/core/cost_storage.py
"""
Cost storage interface for multi-level budget tracking.

Provides storage backends for tracking user-level and global-level usage.
"""

from __future__ import annotations

import logging
from abc import ABC, abstractmethod
from dataclasses import dataclass
from datetime import datetime
from typing import TYPE_CHECKING

from harness.types import UserUsage

if TYPE_CHECKING:
    pass

logger = logging.getLogger(__name__)


@dataclass
class GlobalUsage:
    """Global usage statistics."""

    daily_cost_usd: float = 0.0
    daily_tokens: int = 0
    date: str = ""  # YYYY-MM-DD format


class CostStorage(ABC):
    """Abstract base class for cost storage."""

    @abstractmethod
    def get_user_usage(self, user_id: str) -> UserUsage:
        """Get usage for a user."""
        pass

    @abstractmethod
    def record_user_usage(
        self,
        user_id: str,
        input_tokens: int = 0,
        output_tokens: int = 0,
        request: bool = False,
    ) -> UserUsage:
        """Record usage for a user."""
        pass

    @abstractmethod
    def get_global_usage(self) -> GlobalUsage:
        """Get global usage."""
        pass

    @abstractmethod
    def record_global_usage(
        self,
        cost_usd: float = 0.0,
        tokens: int = 0,
    ) -> GlobalUsage:
        """Record global usage."""
        pass

    @abstractmethod
    def reset_daily(self) -> None:
        """Reset daily counters."""
        pass


class InMemoryCostStorage(CostStorage):
    """
    In-memory cost storage.

    Suitable for single-process applications. Data is lost on restart.

    Example:
        >>> storage = InMemoryCostStorage()
        >>> usage = storage.record_user_usage("user-123", input_tokens=1000)
        >>> print(usage.daily_tokens)
    """

    def __init__(self):
        self._user_usage: dict[str, UserUsage] = {}
        self._global_usage = GlobalUsage()
        self._last_reset_date: str = ""

    def _get_current_date(self) -> str:
        """Get current date string."""
        return datetime.now().strftime("%Y-%m-%d")

    def _get_current_hour(self) -> int:
        """Get current hour."""
        return datetime.now().hour

    def _check_and_reset_daily(self) -> None:
        """Check if we need to reset daily counters."""
        current_date = self._get_current_date()
        if current_date != self._last_reset_date:
            self.reset_daily()
            self._last_reset_date = current_date

    def get_user_usage(self, user_id: str) -> UserUsage:
        """Get usage for a user."""
        self._check_and_reset_daily()

        if user_id not in self._user_usage:
            self._user_usage[user_id] = UserUsage(
                user_id=user_id,
                date=self._get_current_date(),
                hour=self._get_current_hour(),
            )

        return self._user_usage[user_id]

    def record_user_usage(
        self,
        user_id: str,
        input_tokens: int = 0,
        output_tokens: int = 0,
        request: bool = False,
    ) -> UserUsage:
        """Record usage for a user."""
        self._check_and_reset_daily()

        if user_id not in self._user_usage:
            self._user_usage[user_id] = UserUsage(
                user_id=user_id,
                date=self._get_current_date(),
                hour=self._get_current_hour(),
            )

        usage = self._user_usage[user_id]

        # Check if we're in a new hour
        current_hour = self._get_current_hour()
        if current_hour != usage.hour:
            usage.hourly_requests = 0
            usage.hour = current_hour

        usage.daily_tokens += input_tokens + output_tokens
        if request:
            usage.hourly_requests += 1

        return usage

    def get_global_usage(self) -> GlobalUsage:
        """Get global usage."""
        self._check_and_reset_daily()
        return self._global_usage

    def record_global_usage(
        self,
        cost_usd: float = 0.0,
        tokens: int = 0,
    ) -> GlobalUsage:
        """Record global usage."""
        self._check_and_reset_daily()

        self._global_usage.daily_cost_usd += cost_usd
        self._global_usage.daily_tokens += tokens

        if not self._global_usage.date:
            self._global_usage.date = self._get_current_date()

        return self._global_usage

    def reset_daily(self) -> None:
        """Reset daily counters."""
        self._user_usage.clear()
        self._global_usage = GlobalUsage(date=self._get_current_date())
        self._last_reset_date = self._get_current_date()


class SQLiteCostStorage(CostStorage):
    """
    SQLite-based cost storage.

    Provides persistent storage for multi-process applications.

    Example:
        >>> storage = SQLiteCostStorage("~/.harness/costs.db")
        >>> usage = storage.record_user_usage("user-123", input_tokens=1000)
    """

    def __init__(self, db_path: str = ".harness/costs.db"):
        from pathlib import Path

        self.db_path = Path(db_path).expanduser()
        self.db_path.parent.mkdir(parents=True, exist_ok=True)
        self._init_db()

    def _init_db(self) -> None:
        """Initialize database schema."""
        import sqlite3

        conn = sqlite3.connect(self.db_path)
        try:
            conn.execute("""
                CREATE TABLE IF NOT EXISTS user_usage (
                    user_id TEXT,
                    date TEXT,
                    hour INTEGER,
                    daily_tokens INTEGER DEFAULT 0,
                    hourly_requests INTEGER DEFAULT 0,
                    PRIMARY KEY (user_id, date)
                )
            """)
            conn.execute("""
                CREATE TABLE IF NOT EXISTS global_usage (
                    date TEXT PRIMARY KEY,
                    daily_cost_usd REAL DEFAULT 0,
                    daily_tokens INTEGER DEFAULT 0
                )
            """)
            conn.commit()
        finally:
            conn.close()

    def _get_current_date(self) -> str:
        """Get current date string."""
        return datetime.now().strftime("%Y-%m-%d")

    def _get_current_hour(self) -> int:
        """Get current hour."""
        return datetime.now().hour

    def get_user_usage(self, user_id: str) -> UserUsage:
        """Get usage for a user."""
        import sqlite3

        date = self._get_current_date()
        hour = self._get_current_hour()

        conn = sqlite3.connect(self.db_path)
        try:
            cursor = conn.execute(
                "SELECT daily_tokens, hourly_requests, hour "
                "FROM user_usage WHERE user_id = ? AND date = ?",
                (user_id, date),
            )
            row = cursor.fetchone()

            if row:
                stored_hour = row[2]
                # If hour changed, reset hourly counter
                hourly_requests = row[1] if stored_hour == hour else 0
                return UserUsage(
                    user_id=user_id,
                    daily_tokens=row[0],
                    hourly_requests=hourly_requests,
                    date=date,
                    hour=hour,
                )

            return UserUsage(user_id=user_id, date=date, hour=hour)
        finally:
            conn.close()

    def record_user_usage(
        self,
        user_id: str,
        input_tokens: int = 0,
        output_tokens: int = 0,
        request: bool = False,
    ) -> UserUsage:
        """Record usage for a user."""
        import sqlite3

        date = self._get_current_date()
        hour = self._get_current_hour()
        tokens = input_tokens + output_tokens

        conn = sqlite3.connect(self.db_path)
        try:
            # Get current values
            cursor = conn.execute(
                "SELECT daily_tokens, hourly_requests, hour "
                "FROM user_usage WHERE user_id = ? AND date = ?",
                (user_id, date),
            )
            row = cursor.fetchone()

            if row:
                stored_hour = row[2]
                # Reset hourly if hour changed
                hourly_requests = row[1] if stored_hour == hour else 0

                conn.execute(
                    """
                    UPDATE user_usage
                    SET daily_tokens = daily_tokens + ?,
                        hourly_requests = ?,
                        hour = ?
                    WHERE user_id = ? AND date = ?
                """,
                    (
                        tokens,
                        hourly_requests + (1 if request else 0),
                        hour,
                        user_id,
                        date,
                    ),
                )
            else:
                conn.execute(
                    """
                    INSERT INTO user_usage (user_id, date, hour, daily_tokens, hourly_requests)
                    VALUES (?, ?, ?, ?, ?)
                """,
                    (user_id, date, hour, tokens, 1 if request else 0),
                )

            conn.commit()
        finally:
            conn.close()

        return self.get_user_usage(user_id)

    def get_global_usage(self) -> GlobalUsage:
        """Get global usage."""
        import sqlite3

        date = self._get_current_date()

        conn = sqlite3.connect(self.db_path)
        try:
            cursor = conn.execute(
                "SELECT daily_cost_usd, daily_tokens FROM global_usage WHERE date = ?",
                (date,),
            )
            row = cursor.fetchone()

            if row:
                return GlobalUsage(
                    daily_cost_usd=row[0],
                    daily_tokens=row[1],
                    date=date,
                )

            return GlobalUsage(date=date)
        finally:
            conn.close()

    def record_global_usage(
        self,
        cost_usd: float = 0.0,
        tokens: int = 0,
    ) -> GlobalUsage:
        """Record global usage."""
        import sqlite3

        date = self._get_current_date()

        conn = sqlite3.connect(self.db_path)
        try:
            conn.execute(
                """
                INSERT INTO global_usage (date, daily_cost_usd, daily_tokens)
                VALUES (?, ?, ?)
                ON CONFLICT(date) DO UPDATE SET
                    daily_cost_usd = daily_cost_usd + ?,
                    daily_tokens = daily_tokens + ?
            """,
                (date, cost_usd, tokens, cost_usd, tokens),
            )
            conn.commit()
        finally:
            conn.close()

        return self.get_global_usage()

    def reset_daily(self) -> None:
        """Reset daily counters. SQLite version just creates new rows for new day."""
        pass  # SQLite automatically handles new dates
